src/models.py

```diff
@@ -245,6 +259,16 @@ def validation_step(self, batch: Tensor, batch_idx: int):
             on_epoch=True,
             sync_dist=True,
             add_dataloader_idx=True)
+        
+        self.log(
+            name="val_perplexity", 
+            value=perplexity, 
+            prog_bar=True,
+            logger=True, 
+            on_step=False, 
+            on_epoch=True,
+            sync_dist=True,
+            add_dataloader_idx=True)
 
         return loss
```

Comment:
Do we only want to return loss? I think this is fine as this is for validation (and training up above), but it feels a bit odd to track loss and perplexity and only return loss.

Reply:
This function is called by torch lightning, and so it must return loss so that lightning can backpropograte, etc. It does not need to return perplexity, we are simply calculating it and then logging it directly.